flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/FlinkResourceManager.java

```diff
@@ -0,0 +1,796 @@
+		}
+	}
+
+	/**
+	 * Causes the resource manager to announce itself at the new leader JobManager and
+	 * obtains its connection information and currently known TaskManagers.
+	 *
+	 * @param leaderAddress The akka actor URL of the new leader JobManager.
+	 */
+	private void triggerConnectingToJobManager(String leaderAddress) {
+		LOG.info("Trying to associate with JobManager leader " + leaderAddress);
+
+		final Object registerMessage = decorateMessage(new RegisterResourceManager(self()));
+		final Object retryMessage = decorateMessage(new TriggerRegistrationAtJobManager(leaderAddress));
+
+		// send the registration message to the JobManager
+		ActorSelection jobManagerSel = context().actorSelection(leaderAddress);
+		Future<Object> future = Patterns.ask(jobManagerSel, registerMessage, new Timeout(messageTimeout));
+
+		future.onComplete(new OnComplete<Object>() {
+
+			@Override
+			public void onComplete(Throwable failure, Object msg) {
+				if (msg != null) {
+					if (msg instanceof LeaderSessionMessage &&
+						((LeaderSessionMessage) msg).message() instanceof RegisterResourceManagerSuccessful)
+					{
+						self().tell(msg, ActorRef.noSender());
+					}
+					else {
+						LOG.error("Invalid response type to registration at JobManager: {}", msg);
+						self().tell(retryMessage, ActorRef.noSender());
+					}
+				}
+				else {
+					// no success
+					LOG.error("Resource manager could not register at JobManager", failure);
+					self().tell(retryMessage, ActorRef.noSender());
+				}
+			}
+
+		}, context().dispatcher());
+	}
+
+	/**
+	 * This method disassociates from the current leader JobManager. All currently registered
+	 * TaskManagers are put under "awaiting registration".
+	 */
+	private void jobManagerLostLeadership() {
+		if (jobManager != null) {
+			LOG.info("Associated JobManager {} lost leader status", jobManager);
+
+			jobManager = null;
+			leaderSessionID = null;
+
+			infoMessageListeners.clear();
+
+			registeredWorkers.clear();
+		}
+	}
+
+	/**
+	 * Callback when we're informed about a new leading JobManager.
+	 * @param newJobManagerLeader The ActorRef of the new jobManager
+	 * @param workers The existing workers the JobManager has registered.
+	 */
+	private void jobManagerLeaderConnected(
+						ActorRef newJobManagerLeader,
+						Collection<ResourceID> workers) {
+
+		if (jobManager == null) {
+			LOG.info("Resource Manager associating with leading JobManager {} - leader session {}",
+						newJobManagerLeader, leaderSessionID);
+
+			jobManager = newJobManagerLeader;
+
+			// inform the framework that we have updated the leader
+			leaderUpdated();
+
+			if (workers.size() > 0) {
+				LOG.info("Received TaskManagers that were registered at the leader JobManager. " +
+						"Trying to consolidate.");
+
+				// keep track of which TaskManagers are not handled
+				List<ResourceID> toHandle = new ArrayList<>(workers.size());
+				for (ResourceID resourceID : workers) {
+					toHandle.add(resourceID);
+				}
```

Comment:
Just FYI, I was curious to see the effects of arraycopy vs. for loop and tested it with copying an 8 element int array. The for loop variant was consistently 50% slower than the arraycopy variant (on average).